cloud/amazon/ec2_vpc_nat_gateway.py

```diff
@@ -0,0 +1,251 @@
+                    module.fail_json(msg="Failed to create VPC NAT Gateway")
+                elif present_status.get('state') == 'available':
+                    nat_gateway_id = result['NatGateway']['NatGatewayId']
+                    module.exit_json(changed=True, nat_gateway_id=nat_gateway_id)
+            else:
+                time.sleep(5)
+        module.fail_json(msg="Waited too long for VPC NAT Gateway to be created.")
+
+
+def ensure_nat_gateway_absent(module=None):
+    """ Ensure the specified NAT gateway does not exist and call delete if it does """
+    ec2_client = get_ec2_client(module=module)
+    try:
+        gateway_status_list = get_nat_gateway_status_list(ec2_client=ec2_client, module=module)
+        if gateway_status_list[0] and gateway_status_list[0].get('state') in ('pending', 'failed', 'available'):
+            if module.check_mode:
+                module.exit_json(changed=True)
+            else:
+                delete_nat_gateway(ec2_client=ec2_client, module=module)
+        else:
+            module.exit_json(changed=False)
+    except botocore.exceptions.ClientError as ce:
+        if "NatGatewayMalformed" in ce.message:
+            module.fail_json(msg="Invalid NAT Gateway ID")
+        else:
+            module.fail_json(msg=ce.message)
+
+
+def ensure_nat_gateway_present(module=None):
+    """ Ensure NAT gateway with specified parameters exists and call create if it does not """
+    ec2_client = get_ec2_client(module=module)
+    try:
+        gateway_status_list = get_nat_gateway_status_list(ec2_client=ec2_client, module=module)
+        for gateway_status in gateway_status_list:
+            if gateway_status.get('state') == 'available':
+                module.exit_json(changed=False)
+        else:
+            if not module.check_mode:
+                create_nat_gateway(ec2_client=ec2_client, module=module)
+            else:
+                module.exit_json(changed=True)
+    except botocore.exceptions.ClientError as ce:
+        if "NatGatewayLimitExceeded" in ce.message:
+            module.fail_json(msg="The NAT Gateway limit has been exceeded.")
+    except botocore.exceptions.NoCredentialsError:
+        module.fail_json(msg="Unable to locate AWS credentials.")
+    except Exception as ce:
+        module.fail_json(msg=ce.message)
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        state=dict(default='present', choices=['present', 'absent']),
+        subnet_id=dict(required=False, type='str'),
+        allocation_id=dict(required=False, type='str'),
+        nat_gateway_id=dict(required=False, type='str'),
+        wait=dict(required=False, type='bool', default=True),
+        wait_timeout=dict(required=False, default=300)))
+
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+    if module.params['state'] == 'present':
```

Comment:
Consider using the (poorly documented, as far as I can tell) required_if keyword in the module rather than this. I suggest reading [other examples](https://github.com/ansible/ansible-modules-core/search?utf8=%E2%9C%93&q=required_if) and [the source](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/basic.py#L1155) to see how it works.